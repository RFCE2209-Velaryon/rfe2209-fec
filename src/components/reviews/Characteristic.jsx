import React from 'react';

const beforeLabels = {
  Size: 'A Size too Small',
  Width: 'Too Narrow',
  Comfort: 'Uncomfortable',
  Quality: 'Poor',
  Length: 'Runs Short',
  Fit: 'Runs Tight'
};

const afterLabels = {
  Size: 'A Size too Big',
  Width: 'Too Wide',
  Comfort: 'Perfect',
  Quality: 'Perfect',
  Length: 'Runs Long',
  Fit: 'Runs Long'
};

const Characteristic = ({characteristic, value}) => {
  return (
    <div>
      <div>{characteristic}</div>
      <input type="range" value={value * 100} min="0" max="500" disabled></input>
      <div className='rangeLabelContainer'>
        <div>{beforeLabels[characteristic]}</div>
        <div>{afterLabels[characteristic]}</div>
      </div>
    </div>
  )
}

export default Characteristic;