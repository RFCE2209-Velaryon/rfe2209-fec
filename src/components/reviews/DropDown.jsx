import React from 'react';

const DropDown = ({choices, cb}) => {
  const [isOpen, setOpen] = React.useState(false);
  const [options, setOptions] = React.useState(choices);
  const [selectedItem, setSelectedItem] = React.useState(choices[0].value);

  function handleItemClick(id) {
    setSelectedItem(id);
    cb(id);
  }

  return (
    <div className='dropdown'>
      <div className='dropdown-header' onClick={() => setOpen(!isOpen)}>
        <h1>{options.find(option => option.value == selectedItem).label}</h1>
        <i className={`icon ${isOpen && "open"}`}></i>
      </div>
      <div className={`dropdown-body ${isOpen && 'open'}`}>
        {options.map(option => (
          <div className="dropdown-item" onClick={e => handleItemClick(e.target.id)} id={option.value} key={option.value}>
            <span className={`dropdown-item-dot ${option.value == selectedItem && 'selected'}`}>â€¢ </span>
            {option.label}
          </div>
        ))}
      </div>
    </div>
  )
}

export default DropDown;