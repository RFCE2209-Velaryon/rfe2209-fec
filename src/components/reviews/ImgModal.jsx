import React from 'react';

const ImgModal = ({photo}) => {
  const [open, setOpen] = React.useState(false);

  function onToggle() {
    setOpen(!open);
  }

  return (
    <div>
      <img onClick={onToggle} className='modalImageThumbnail' src={photo.url} />
      {open && <div onClick={onToggle} className='grayModalBackground'>
                <img className='modalImage' src={photo.url} />
                <div className='closeModalImage'>X</div>
               </div>
               }
    </div>
  )
}

export default ImgModal;