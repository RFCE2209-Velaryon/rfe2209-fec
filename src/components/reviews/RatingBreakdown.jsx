import React from 'react';

const RatingBreakdown = ({ratings, filters, setFilters}) => {
  function getRatingsOfMostCommonRating() {
    let largest = 0;
    for(const key in ratings) {
      if(Number(ratings[key]) > largest) {
        largest = ratings[key];
      }
    }
    return largest;
  }

  function filterByRating(e) {
    if(!filters[Number(e.target.innerHTML.substr(0, 1))]) {
      let newFilters = Object.assign({}, filters);
      newFilters[Number(e.target.innerHTML.substr(0, 1))] = true;
      setFilters(newFilters);
    }
  }

  function filtered() {
    let filtersApplied = false;
    Object.keys(filters).forEach(key => {
      filtersApplied = filters[key] || filtersApplied;
    });
    return filtersApplied;
  }

  return (
    <div className='ratingBreakdownList'>
      {[5, 4, 3, 2, 1].map(index => {
        return (
        <div className='ratingBreakdown' key={index}>
          <h4 className='ratingBreakdownLabel' onClick={filterByRating}>{`${index} stars`}</h4>
          <div className="progress-bar">
            <div className="progress-bar-fill" style={{width: `${(ratings[index]/getRatingsOfMostCommonRating())*100}%`}}></div>
          </div>
        </div>
        )
      })}
    </div>
  )
}

export default RatingBreakdown;