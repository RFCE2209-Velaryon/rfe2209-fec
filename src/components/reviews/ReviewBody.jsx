import React from 'react';

const ReviewBody = ({body}) => {
  if(body.length > 250) {
    const [hidden, setHidden] = React.useState(true);
    return (
      <div>
        <h4>{hidden ? body.slice(0, 250) + '...' : body}</h4>
        <h4 onClick={() => setHidden(!hidden)}>{hidden ? 'Show More' : 'Show Less'}</h4>
      </div>

    )
  } else {
    return (
      <h4>{body}</h4>
    )
  }
}

export default ReviewBody;