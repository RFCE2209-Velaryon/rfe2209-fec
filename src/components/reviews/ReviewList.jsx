import React from 'react';
import Review from './review.jsx';

const ReviewList = ({reviews}) => {
  return (
    <div className="reviewList">
      {reviews.map((review, i) => {
        return <Review key={i} review={review}/>;
      })}
    </div>
  )
}

export default ReviewList;