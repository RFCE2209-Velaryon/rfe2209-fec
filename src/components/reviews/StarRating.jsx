import React from 'react';

const StarRating = ({initialRating, readOnly, cb}) => {


  const [rating, setRating] = React.useState(initialRating);
  const [hover, setHover] = React.useState(0);

  return (
    <div className="star-rating">
      {[...Array(5)].map((star, index) => {
        index += 1;
        return (
          <button
            type="button"
            key={index}
            className={index <= (hover || rating) ? "on" : "off"}
            onClick={() => {
              if(!readOnly) {
                setRating(index);
                if(cb) {
                  cb(index);
                }
              }
            }}
            onMouseEnter={() => {
              if(!readOnly) {
                setHover(index);
              }
            }}
            onMouseLeave={() => {
              if(!readOnly) {
                setHover(rating);
              }
            }}
          >
            <span className="star">&#9733;</span>
          </button>
        );
      })}
    </div>
  );
};

export default StarRating;