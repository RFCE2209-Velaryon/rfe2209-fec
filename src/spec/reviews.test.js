import renderer from 'react-test-renderer';
import React from 'react';

import Reviews from '../components/reviews/reviews.jsx';
import Characteristic from '../components/reviews/Characteristic.jsx';
import ImgModal from '../components/reviews/ImgModal.jsx';
import RatingBreakdown from '../components/reviews/RatingBreakdown.jsx';
import Review from '../components/reviews/review.jsx';
import ReviewBody from '../components/reviews/ReviewBody.jsx';
import ReviewList from '../components/reviews/ReviewList.jsx';
import StarRating from '../components/reviews/StarRating.jsx';
import DropDown from '../components/reviews/DropDown.jsx';



describe("Renders Components Correctly" ,() => {

  // Test 1
  test('Reviews', () => {
    const product = {
      "id": 37311,
      "campus": "hr-rfe",
      "name": "Camo Onesie",
      "slogan": "Blend in to your crowd",
      "description": "The So Fatigues will wake you up and fit you in. This high energy camo will have you blending in to even the wildest surroundings.",
      "category": "Jackets",
      "default_price": "140.00",
      "created_at": "2021-08-13T14:37:33.145Z",
      "updated_at": "2021-08-13T14:37:33.145Z"
    }
    const tree = renderer
      .create(<Reviews product={product}/>)
      .toJSON();
      expect(tree).toMatchSnapshot();
   });
   test('Characteristic', () => {
    const tree = renderer
      .create(<Characteristic characteristic='Fit' value={3.0972222222222222} />)
      .toJSON();
      expect(tree).toMatchSnapshot();
   });
   test('ImgModal', () => {
    const tree = renderer
      .create(<ImgModal photo={{"id":2456558,"url":"http://res.cloudinary.com/de2i2agjs/image/upload/v1666886937/r7k7es27yklyhlqju6ls.webp"}}/>)
      .toJSON();
      expect(tree).toMatchSnapshot();
   });
   test('RatingBreakdown', () => {
    const tree = renderer
      .create(<RatingBreakdown ratings={{"1":"50","2":"29","3":"82","4":"114","5":"291"}} filters={{"1":false,"2":false,"3":false,"4":false,"5":false}} />)
      .toJSON();
      expect(tree).toMatchSnapshot();
   });
   test('Review', () => {
    const tree = renderer
      .create(<Review review={{"review_id":1277183,"rating":2,"summary":"do not report","recommend":true,"response":null,"body":"there weren't any 2 star reviews so it made my filter look like it was broken. Help me help you","date":"2022-10-25T00:00:00.000Z","reviewer_name":"FrodoSwaggins","helpfulness":20,"photos":[{"id":2456464,"url":"http://res.cloudinary.com/dqmnjwd2c/image/upload/v1666661253/departed_ete3as.jpg"}]}} />)
      .toJSON();
      expect(tree).toMatchSnapshot();
   });
   test('ReviewBody', () => {
    const tree = renderer
      .create(<ReviewBody body='This is the body text of a review. It is indeed, some text, which goes, into a review.'/>)
      .toJSON();
      expect(tree).toMatchSnapshot();
   });
   test('ReviewList', () => {
    const tree = renderer
      .create(<ReviewList reviews={[{"review_id":1277183,"rating":2,"summary":"do not report","recommend":true,"response":null,"body":"there weren't any 2 star reviews so it made my filter look like it was broken. Help me help you","date":"2022-10-25T00:00:00.000Z","reviewer_name":"FrodoSwaggins","helpfulness":20,"photos":[{"id":2456464,"url":"http://res.cloudinary.com/dqmnjwd2c/image/upload/v1666661253/departed_ete3as.jpg"}]},{"review_id":1277038,"rating":1,"summary":"There and back again","recommend":false,"response":null,"body":"A \"no\" on recommendation should change the metadata I get back","date":"2022-10-22T00:00:00.000Z","reviewer_name":"frodo","helpfulness":17,"photos":[]}]} />)
      .toJSON();
      expect(tree).toMatchSnapshot();
   });
   test('StarRating', () => {
    const tree = renderer
      .create(<StarRating initialRating={3} />)
      .toJSON();
      expect(tree).toMatchSnapshot();
   });
   test('DropDown', () => {
    const tree = renderer
      .create(<DropDown choices={[{"label":"Relevance","value":"relevant"},{"label":"Helpfulness","value":"helpful"},{"label":"Date","value":"newest"}]} />)
      .toJSON();
      expect(tree).toMatchSnapshot();
   });
});

// describe("Renders the text props given" ,() => {
//   test('ReviewBody', () => {
//     const tree = renderer.create(<ReviewBody body='This is the body text of a review. It is indeed, some text, which goes, into a review.' />);
//     tree.root.find(findInChildren(node =>
//       typeof node === 'string' &&
//       node === 'text of a review.'
//     ));
//   });
// });


// // helper functions
// function findInChildren (predicate) {
//   return testInstance => {
//     const children = testInstance.children
//     return Array.isArray(children)
//       ? children.some(predicate)
//       : predicate(children)
//   }
// }