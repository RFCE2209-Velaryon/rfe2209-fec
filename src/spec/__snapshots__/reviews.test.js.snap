// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Reviews Component renders correctly 1`] = `
<h1>
  Review Component
</h1>
`;
